grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy

```diff
@@ -341,30 +342,32 @@ class StockMovementApiController {
     }
 
     def exportPickListItems() {
+        Boolean isTemplate = params.boolean("template", false)
+
         List<PickPageItem> pickPageItems = stockMovementService.getPickPageItems(params.id, null, null )
-        List<PicklistItem> picklistItems = pickPageItems.inject([]) { result, pickPageItem ->
-            result.addAll(pickPageItem.picklistItems)
-            result
-        }
+        List<PicklistItem> picklistItems = pickPageItems.picklistItems?.flatten()
         // We need to create at least one row to ensure an empty template
         if (picklistItems?.empty) {
             picklistItems.add(new PicklistItem())
         }
 
-        def lineItems = picklistItems.collect {
+        List lineItems = picklistItems.collect {
             [
-                    "${g.message(code: 'default.id.label')}": it?.requisitionItem?.id ?: "",
-                    "${g.message(code: 'product.productCode.label')}": it?.requisitionItem?.product?.productCode ?: "",
-                    "${g.message(code: 'product.name.label')}": it?.requisitionItem?.product?.name ?: "",
-                    "${g.message(code: 'inventoryItem.lotNumber.label')}": it?.inventoryItem?.lotNumber ?: "",
-                    "${g.message(code: 'inventoryItem.expirationDate.label')}": it?.inventoryItem?.expirationDate ? it.inventoryItem.expirationDate.format(Constants.EXPIRATION_DATE_FORMAT) : "",
-                    "${g.message(code: 'inventoryItem.binLocation.label')}": it?.binLocation?.name ?: "",
-                    "${g.message(code: 'default.quantity.label')}": it?.quantity ?: "",
+                    id: it?.requisitionItem?.id,
+                    code: it?.requisitionItem?.product?.productCode,
+                    name: it?.requisitionItem?.product?.name,
+                    lot: isTemplate ? null : it?.inventoryItem?.lotNumber,
+                    expiration: isTemplate ? null : it.inventoryItem.expirationDate?.format(Constants.EXPIRATION_DATE_FORMAT),
+                    binLocation: isTemplate ? null : it?.binLocation?.name,
+                    quantity: isTemplate ? it?.requisitionItem?.quantity : it?.quantity,
             ]
         }
-        String csv = dataService.generateCsv(lineItems)
-        response.setHeader("Content-disposition", "attachment; filename=\"StockMovementItems-${params.id}.csv\"")
-        render(contentType: "text/csv", text: csv.toString(), encoding: "UTF-8")
+
+        response.contentType = "application/vnd.ms-excel"
+        response.setHeader("Content-disposition", "attachment; filename=\"StockMovementItems-${params.id}.xls\"")
+        PickListItemExcelExporter pickListItemExcelExporter = new PickListItemExcelExporter(lineItems)
```

Comment:
As I mentioned on the call, let's not try to do anything new here. It was a good attempt, but I would like to make sure that new things are linked to a design and design review process, which will be handled as part of the discovery tickets and design sprints that Evan and I have mentioned in the past few weeks. If there's room to do a discovery and implementation in the same sprint, then great. But I suspect we're going to want to do these as separate activities across multiple sprints.

So for now, let's follow the current convention and add a generateExcel() method to dataService and we can work on OBIG-4 separately.

https://pihemr.atlassian.net/browse/OBIG-4

For this implementation, we should try to keep the previous code as-is (maybe move the main data retrieveal to a service) and just add in a way to switch on the rendering format.

This pseudocode - write this however you see fit (conditional, switch, etc) I just wanted to explain what I was thinking. I left out the code for the empty template case, so I'll let you figure out whether that should be included.
```
// Retrieve (and myabe validate) the stock movement for the given ID i.e. whether there is a stock movement with that id
def stockMovement = stockMovementService.getStockMovement(params.id)

// Keep the data building block from the previous version, but move to a service method.
def lineItems = picklistOrStockMovementService.getMeSomeData(stockMovement.picklist.id)

// FIXME We should eventually also handle content negotiation via file extension, request headers 
// and request parameters
if (!params.format || params.format == "csv") { 
    String csv = dataService.generateCsv(lineItems)
    response.setHeader("Content-disposition", "attachment; filename=\"StockMovementItems-${params.id}.csv\"")
    render(contentType: "text/csv", text: data.toString(), encoding: "UTF-8")
    return
} 
else if (params.format == "xls") {
    Workbook workbook = dataService.generateExcel(lineItems) 
    response.contentType = "application/vnd.ms-excel"
    response.setHeader("Content-disposition", "attachment; filename=\"StockMovementItems-${params.id}.xls\"")
    workbook.write(response.outputStream)
    response.outputStream.flush()
}
else { 
    throw new IllegalFormatException("Unable to determine the proper rendering format for request for format ${params.format}")
}

```